### PR TITLE
Create service using template from an external repository

### LINKED ISSUE #4099: Feature Request: Custom Template
# This is a Feature Proposal

## Description

The standard templates are great generic boilerplates. However, I have to make the same modifications everytime I start a new project. I think it would be great to be able to create and specify my own custom template for use with the service create command.

### PR DESCRIPTION
**Pull request by kbariotis**
## What did you implement:

Closes #4099

* Abstract the download template from external repo function
* Add support for BitBucket also
* Utilize it at `create` command and add the option to pull template from repo

## How did you implement it:

`install` command would accept a repository URL and download it. I have moved this function into its own utility in order to be shared by the `create` command too.

## How can we verify it:

`sls create --template-url https://github.com/serverless/examples/tree/master/aws-node-rest-api-mongodb`

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by RafalWilinski** [comment]
@kbariotis Thanks for this PR 🙌

Is it ready for review?

File: lib/utils/fs/dirExistsSync.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const fse = require('./fse');
+
+function dirExistsSync(dirPath) {
```
Comment (RafalWilinski) [inline lib/utils/fs/dirExistsSync.js]: Possible code duplication: https://github.com/serverless/serverless/blob/master/lib/classes/Utils.js#L27

File: lib/utils/renameService.test.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    const serviceFile = path.join(servicePath, 'serverless.yml');
+
+    serverless.utils.writeFileSync(packageFile, { name: defaultServiceName });
+    fse.writeFileSync(serviceFile, defaultServiceYml);
+
+    renameService(newServiceName, servicePath);
+    const serviceYml = fse.readFileSync(serviceFile, 'utf-8');
+    const packageJson = serverless.utils.readFileSync(packageFile);
+    expect(serviceYml).to.equal(newServiceYml);
+    expect(packageJson.name).to.equal(newServiceName);
```
Comment (RafalWilinski) [inline lib/utils/renameService.test.js]: I think you might need to remove those files after tests

File: lib/utils/renameService.test.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      'service: new-service-name\n\nprovider:\n  name: aws\n';
+
+    const defaultServiceName = 'service-name';
+    const newServiceName = 'new-service-name';
+
+    const packageFile = path.join(servicePath, 'package.json');
+    const serviceFile = path.join(servicePath, 'serverless.yml');
+
+    serverless.utils.writeFileSync(packageFile, { name: defaultServiceName });
+    fse.writeFileSync(serviceFile, defaultServiceYml);
```
Comment (RafalWilinski) [inline lib/utils/renameService.test.js]: I think you might need to remove those files after tests

Comment (kbariotis) [reply lib/utils/renameService.test.js]: I didnt see other tests removing the temp directory. Do you have an example of another test?

**Comment by kbariotis** [comment]
Thank you @RafalWilinski! PR is ready to review, though I have to add documentation. :) I will soon.

**Comment by RafalWilinski** [comment]
Thanks for adding docs @kbariotis, looking forward to merge this PR! 🙌

Will review that in the upcoming days.

Also cc'ing @horike37 to double check

File: docs/providers/aws/cli-reference/create.md [hunk 4]
```diff
@@ -26,8 +26,15 @@ serverless create --template aws-nodejs
...
+**Create service in new folder using a custom template:**
+
+```bash
+serverless create --template-url https://github.com/serverless/serverless/tree/master/lib/plugins/create/templates/aws-nodejs --path myService
+```
+
 ## Options
-- `--template` or `-t` The name of one of the available templates. **Required**.
+- `--template` or `-t` The name of one of the available templates. **Required if --template-url is not present**.
+- `--template-url` or `-t` The name of one of the available templates. **Required if --template is not present**.
```
Comment (horike37) [inline docs/providers/aws/cli-reference/create.md]: -t --> -u in the short version.

File: lib/utils/renameService.test.js (hunk 3 above)
Comment (HyperBrain) [reply lib/utils/renameService.test.js]: Woudn't it be even better to stub the file access methods in the implementation under testing? Having unit tests that rely on any physical filesystem seems odd (we do not want to test `fs` or `fs-extra` - we should assume these work).
PS: I know that this is also the case in other areas of the unit tests, but that does not mean to continue with that habit 😄

**Comment by kbariotis** [comment]
@horike37 im sorry but those conflicts are really frustrating me. Is there a chance that you have a better view on how they should be resolved?

**Comment by kbariotis** [comment]
@horike37 this is finally ready. Thank you for your patience. :)

**Comment by RafalWilinski** [comment]
Thanks for taking your time to ship this feature and fulfull our requests, it's awesome! ❤️  

Reviewing it right now.

File: lib/utils/downloadTemplateFromRepo.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  // download service
+  return download(
+    repoInformation.downloadUrl,
+    downloadServicePath,
+    { timeout: 30000, extract: true, strip: 1, mode: '755' }
+  ).then(() => {
+    // if it's a directory inside of git
+    if (repoInformation.isSubdirectory) {
+      const directory = path.join(downloadServicePath, repoInformation.pathToDirectory);
+      copyDirContentsSync(directory, servicePath);
```
Comment (RafalWilinski) [inline lib/utils/downloadTemplateFromRepo.js]: Usually, we try to get rid of `sync` calls in favor of `async` ones but I don't think that's an issue here. It's not an integral part of framework invoked very frequently.

**Comment by kbariotis** [comment]
@RafalWilinski GitLab differs from the others in that it supports sub-organizations. So it may need a bit more time to find the right formula that will extract the information we need from the URL.

I'm trying to find out about the tests now. Hope I will make them run soon.

**Comment by RafalWilinski** [comment]
Actually, tests fail because of other, not related to this PR issue. This shouldn't be your concern, unless you'd like to help investigating it 😄 https://github.com/serverless/serverless/issues/4350

**Comment by kumarryogeshh** [comment]
Thank you @kbariotis for the PR. Great work mate!
Waiting for it to get merged.

**Comment by RafalWilinski** [comment]
@horike37 fixed https://github.com/serverless/serverless/issues/4350 problem so nothing blocks us! 

Thanks again @kbariotis 💯 

Merging...

**Comment by hadolphs** [comment]
Guys, was this released? we are looking at the code in version 1.25 and in the create.js is missing the template-path functionality... is this something is targeted for the next version?

**Comment by horike37** [comment]
This PR is about `--template-url` functionality, not `--template-path`.
`--template-path` functionality has been merged in the master branch and then will be released in v1.26.
https://github.com/serverless/serverless/pull/4576

**Comment by hadolphs** [comment]
Thanks, actually now that you mention it, I see here that this is supposed to work with bitbucket... but we haven't been able to make it work with it... how did you guys test it to make it work with a bitbucket repo? does it have to be a public repo?

**Comment by garethmcc** [comment]
I cannot get this to work either using either install or create. Can someone perhaps provide some clarity?

**Comment by garethmcc** [comment]
I have discovered that this only works on public repositories.

**Comment by iksnae** [comment]
is there a way to make this work with private repositories?

**Pull request merged**